File: 3614-process-string-with-special-operations-ii/3614-process-string-with-special-operations-ii.cpp
class Solution {
public:
    char processStr(string s, long long k) {
        long long len = 0;
        for (char c : s) {
            if (c == '*') {
                len = max(0LL, len - 1);
            } else if (c == '#') {
                len *= 2;
            } else if (c != '%') {
                len++;
            }
        }
        if (k >= len)return '.';
        for (int i = (int)s.size() - 1; i >= 0; i--) {
            char c = s[i];
            if (c == '*') {
                len++;
            }else if (c == '#') {
                long long half = len / 2;
                if (k >= half) {
                    k -= half;
                }
                len = half;
            }else if (c == '%') {
                k = len - 1 - k;
            }else {
                if (k == len - 1)return c;
                len--;
            }
        }
        return '.';
    }
};

File: 3614-process-string-with-special-operations-ii/3614-process-string-with-special-operations-ii.py
class Solution:
    def processStr(self, s: str, k: int) -> str:
        length = 0
        for c in s:
            if c == '*':
                length = max(0, length - 1)
            elif c == '#':
                length *= 2
            elif c != '%':
                length += 1
        if k >= length:
            return '.'
        for i in range(len(s) - 1, -1, -1):
            c = s[i]
            if c == '*':
                length += 1
            elif c == '#':
                half = length // 2
                if k >= half:
                    k -= half
                length = half
            elif c == '%':
                k = length - 1 - k
            else:
                if k == length - 1:
                    return c
                length -= 1
        return '.'